### ISSUE TITLE
Path rewriting corrupts files when building on Windows Subsytem for Linux

### ISSUE THREAD

**Issue by steamelephant** [conda/conda-build#2598]
The conda/cookiecutter-conda-python project builds a command line script with a hard-coded path. The rewritten version of the file included in the package starts with the intended content but repeats the end of the file. The file is generated properly when installed by setup.py and when installed in Windows directly and in various more normal Linux installations so I'm hypothesizing some odd ineraction between mmap and the abstracted file system.

This is an edge case but it would be very convenient to be able to build in WSL.  

```python
#!/opt/anaconda1anaconda2anaconda3/bin/python`
# EASY-INSTALL-ENTRY-SCRIPT: 'abc==0.1.0','console_scripts','abc'
__requires__ = 'abc==0.1.0'
import re
import sys
from pkg_resources import load_entry_point

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(
        load_entry_point('abc==0.1.0', 'console_scripts', 'abc')()
    )
pkg_resources import load_entry_point

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(
        load_entry_point('abc==0.1.0', 'console_scripts', 'abc')()
    )
```
Reducing the package name length makes the duplicate tail start earlier:

```python
#!/opt/anaconda1anaconda2anaconda3/bin/python
# EASY-INSTALL-ENTRY-SCRIPT: 'ab==0.1.0','console_scripts','ab'
__requires__ = 'ab==0.1.0'
import re
import sys
from pkg_resources import load_entry_point

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(
        load_entry_point('ab==0.1.0', 'console_scripts', 'ab')()
    )
m pkg_resources import load_entry_point

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(
        load_entry_point('ab==0.1.0', 'console_scripts', 'ab')()
    )
```

**Comment by msarahan** [comment]
If there's something we can do about it, we'll be happy to, but WSL stuff can get pretty far over my head pretty quickly.  If you can isolate this any further, please let us know.

**Comment by steamelephant** [comment]
I've looked a bit deeper. The problem is that on WSL if you try to overwrite a long 'mmap'ed file with a short file you only overwrite the beginning of the file. This is fixed by closing the mmap before writing to the file so instead of:
```
rewrite_file_with_new_prefix(path, mm[:], prefix_bytes, prefix_placeholder_bytes)
mm.close()
fi.close()
fi = open(path, 'rb+')
```
the `build.py` did:
```
data = mm[:]
mm.close()
fi.close()
rewrite_file_with_new_prefix(path, data, prefix_bytes, prefix_placeholder_bytes)
fi = open(path, 'rb+')
```
it ought to work on WSL. 

Alternatively if there were a guarantee that the placeholder prefixe were longer than the prefix that would work too. 

There are current problems with the C mmap library on WSL so I don't hold much hope of it being fixed via the mmap library very soon.

**Comment by msarahan** [comment]
Fixed in #2655; closing
